---
title: "feature_selection_all"
output: html_document
setwd("~/Documents/Github/chemotherapy_response_TNBC/publish")
---

```{r libraries}
library(tidyverse)
library(ggplot2)
library(kableExtra)
library(gplots)
library(reshape2)
library(plyr)
library(stringr)
library(dplyr)
library(tidyr)
library(WGCNA)
library(caret)
library(xgboost)
library(ModelMetrics)
library(rsm)
library(pROC)

human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")
mouse_to_human = function(x){
  mouse_to_human_list<- getLDS(attributes = c("mgi_symbol"), filters = "mgi_symbol", values = x , mart = mouse, attributesL = c("hgnc_symbol"), martL = human, uniqueRows=T)
  return(mouse_to_human_list)}

mouse_to_human_df = read.table("input/mouse_to_human_df.txt")

```

```{r load the response and expression data}
response = read.table("input/pdx_cohort_response.txt", header = TRUE)
response$xeno = gsub("-","_",response$xeno)
pdx_response = as.character(unique(response$xeno))

#Read in the RNA-seq data (Murine and Human)
#log2(1000 * Upper Quartile normalized counts + 1).
xenome_human_RSEM = read.csv("input/PDX_RNAseq_UQCounts_log2_Human.csv",header = TRUE)
rownames(xenome_human_RSEM) = xenome_human_RSEM$X
xenome_mouse_RSEM = read.csv("input/PDX_RNAseq_UQCounts_log2_Mouse.csv")
rownames(xenome_mouse_RSEM) = xenome_mouse_RSEM$X

#Keep only the PDXs with response 
xenome_human_RSEM = xenome_human_RSEM[,colnames(xenome_human_RSEM) %in% pdx_response]
xenome_human_RSEM = data.frame(t(xenome_human_RSEM))
xenome_mouse_RSEM = xenome_mouse_RSEM[,colnames(xenome_mouse_RSEM) %in% pdx_response]
xenome_mouse_RSEM = data.frame(t(xenome_mouse_RSEM))
```

```{r make the response dfs}
response_qual = response
response_qual = response_qual[,c(1,2,7)]
response = response[,c(1,2,4)]

#Subset the response table 
carbo_quant = response[response$tx == "Carbo",]
doc20_quant = response[response$tx == "Doc20",]
doc30_quant = response[response$tx == "Doc30",]
```

```{r diff response}
carbo_quant_score = carbo_quant
colnames(carbo_quant_score)[3] = "Carbo_logfc_estimate"
carbo_quant_score$tx = NULL

doc20_quant_score = doc20_quant
colnames(doc20_quant_score)[3] = "Doc20_logfc_estimate"
doc20_quant_score$tx = NULL

doc30_quant_score = doc30_quant
colnames(doc30_quant_score)[3] = "Doc30_logfc_estimate"
doc30_quant_score$tx = NULL


carbo_qual = response_qual[response_qual$tx == "Carbo",]
doc20_qual = response_qual[response_qual$tx == "Doc20",]
doc30_qual = response_qual[response_qual$tx == "Doc30",]


carbo_qual_score = carbo_qual
colnames(carbo_qual_score)[3] = "Carbo_qual"
carbo_qual_score$tx = NULL

doc20_qual_score = doc20_qual
colnames(doc20_qual_score)[3] = "Doc20_qual"
doc20_qual_score$tx = NULL

doc30_qual_score = doc30_qual
colnames(doc30_qual_score)[3] = "Doc30_qual"
doc30_qual_score$tx = NULL
```

#CTD
```{r load the CTD info (generated by the CTD models file)}
load("input/ctd_info_genes_glm.RData")
ctd_carbo_MAE = carbo_MAE
ctd_carbo_RMSE = carbo_glm_RMSE
ctd_doc_MAE = doc20_MAE
ctd_doc_RMSE = doc20_glm_RMSE
ctd_fit_CRPR = fit_CRPR
ctd_fit_CR = fit_CR
rm(carbo_glm_RMSE,carbo_MAE,doc20_MAE,doc20_glm_RMSE,fit_CR,fit_CRPR)
```

```{r read correlation genes}
#Top 20 most correlated genes
correlation_genes_human_carbo = read.table("input/correlation_genes_human_carbo.txt")
correlation_genes_human_carbo = correlation_genes_human_carbo$x
correlation_genes_human_doc = read.table("input/correlation_genes_human_doc.txt")
correlation_genes_human_doc = correlation_genes_human_doc$x

correlation_genes_human_mouse_carbo = read.table("input/correlation_genes_human_mouse_carbo.txt")
correlation_genes_human_mouse_carbo = correlation_genes_human_mouse_carbo$x
correlation_genes_human_mouse_doc = read.table("input/correlation_genes_human_mouse_doc.txt")
correlation_genes_human_mouse_doc = correlation_genes_human_mouse_doc$x

```

```{r pseudo-bulk cor carbo}
xenome_human_RSEM_cor = xenome_human_RSEM[,colnames(xenome_human_RSEM) %in% correlation_genes_human_carbo]
xenome_human_RSEM_cor$pdx = rownames(xenome_human_RSEM_cor)
xenome_human_RSEM_cor_melt = melt(xenome_human_RSEM_cor)

xenome_mouse_RSEM_cor = xenome_mouse_RSEM[,colnames(xenome_mouse_RSEM) %in% correlation_genes_human_mouse_carbo]
#Change colnames to human homologs 
colnames_cor_mouse = mouse_to_human_df[mouse_to_human_df$genes %in% correlation_genes_human_mouse_carbo,]
colnames_cor_mouse_human = as.character(colnames_cor_mouse$HGNC.symbol)
colnames_cor_mouse$genes = as.character(colnames_cor_mouse$genes)
xenome_mouse_RSEM_cor = xenome_mouse_RSEM[,colnames_cor_mouse$genes]
colnames(xenome_mouse_RSEM_cor) = colnames_cor_mouse$HGNC.symbol


xenome_mouse_RSEM_cor$pdx = rownames(xenome_mouse_RSEM_cor)
xenome_mouse_RSEM_cor_melt = melt(xenome_mouse_RSEM_cor)
xenome_human_mouse_RSEM_cor_melt = rbind(xenome_human_RSEM_cor_melt,xenome_mouse_RSEM_cor_melt)
xenome_human_mouse_RSEM_cor_melt = aggregate(. ~pdx+variable, data=xenome_human_mouse_RSEM_cor_melt, sum, na.rm=TRUE)
xenome_human_mouse_RSEM_cor = recast(xenome_human_mouse_RSEM_cor_melt, pdx ~ variable)
rownames(xenome_human_mouse_RSEM_cor) = xenome_human_mouse_RSEM_cor$pdx
xenome_human_mouse_RSEM_cor$pdx = NULL 
```

```{r pseudo bulk cor doc}
xenome_human_RSEM_cor_doc = xenome_human_RSEM[,colnames(xenome_human_RSEM) %in% correlation_genes_human_doc]
xenome_human_RSEM_cor_doc$pdx = rownames(xenome_human_RSEM_cor_doc)
xenome_human_RSEM_cor_doc_melt = melt(xenome_human_RSEM_cor_doc)

xenome_mouse_RSEM_cor_doc = xenome_mouse_RSEM[,colnames(xenome_mouse_RSEM) %in% correlation_genes_human_mouse_doc]
colnames_cor_doc_mouse = mouse_to_human_df[mouse_to_human_df$genes %in% correlation_genes_human_mouse_doc,]
colnames_cor_doc_mouse_human = as.character(colnames_cor_doc_mouse$HGNC.symbol)
colnames_cor_doc_mouse$genes = as.character(colnames_cor_doc_mouse$genes)
doc_xenome_mouse = intersect(colnames(xenome_mouse_RSEM),colnames_cor_doc_mouse$genes)
colnames_cor_doc_mouse  = colnames_cor_doc_mouse[colnames_cor_doc_mouse$genes %in% doc_xenome_mouse, ]
xenome_mouse_RSEM_cor_doc = xenome_mouse_RSEM[,colnames_cor_doc_mouse$genes]
colnames(xenome_mouse_RSEM_cor_doc) = colnames_cor_doc_mouse$HGNC.symbol


xenome_mouse_RSEM_cor_doc$pdx = rownames(xenome_mouse_RSEM_cor_doc)
xenome_mouse_RSEM_cor_doc_melt = melt(xenome_mouse_RSEM_cor_doc)
xenome_human_mouse_RSEM_cor_doc_melt = rbind(xenome_human_RSEM_cor_doc_melt,xenome_mouse_RSEM_cor_doc_melt)
xenome_human_mouse_RSEM_cor_doc_melt = aggregate(. ~pdx+variable, data=xenome_human_mouse_RSEM_cor_doc_melt, sum, na.rm=TRUE)
xenome_human_mouse_RSEM_cor_doc = recast(xenome_human_mouse_RSEM_cor_doc_melt, pdx ~ variable)
rownames(xenome_human_mouse_RSEM_cor_doc) = xenome_human_mouse_RSEM_cor_doc$pdx
xenome_human_mouse_RSEM_cor_doc$pdx = NULL 
```

```{r  models carbo cor}
xenome_human_mouse_RSEM_cor$xeno = rownames(xenome_human_mouse_RSEM_cor)
score_df_all_carbo_cor = merge(xenome_human_mouse_RSEM_cor,carbo_quant_score, by = "xeno")
rownames(score_df_all_carbo_cor) = score_df_all_carbo_cor$xeno
score_df_all_carbo_cor$xeno = NULL
colnames(score_df_all_carbo_cor)[21] = "response"

score_df_all_carbo_cor_qual = merge(xenome_human_mouse_RSEM_cor,carbo_qual_score, by = "xeno")
rownames(score_df_all_carbo_cor_qual) = score_df_all_carbo_cor_qual$xeno
score_df_all_carbo_cor_qual$xeno = NULL
colnames(score_df_all_carbo_cor_qual)[21] = "response"

split_10_carbo_cor = read.table("input/split_10_carbo_quant.txt")

carbo_cor_glm = list()
carbo_cor_glm_pred = list()
carbo_cor_glm_RMSE = list()
carbo_cor_MAE = list()
carbo_cor_plot = list()

for (i in 1:10){
  split = split_10_carbo_cor[,i]
  score_df_all_carbo_cor_train = score_df_all_carbo_cor[split,]
  score_df_all_carbo_cor_test = score_df_all_carbo_cor[-split,]
  glm_model = glm(response ~ ., data = score_df_all_carbo_cor_train,maxit = 500)
  carbo_cor_glm[[i]] = glm
  predictions <- glm_model %>% predict(score_df_all_carbo_cor_test, type = "response")
  df_pred_obs = cbind(predictions,score_df_all_carbo_cor_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  carbo_cor_glm_pred[[i]] = df_pred_obs
  RMSE = RMSE(predictions,score_df_all_carbo_cor_test$response)
  carbo_cor_glm_RMSE[[i]] = RMSE
  print(RMSE)
  MAE = MAE(predictions,score_df_all_carbo_cor_test$response)
  carbo_cor_MAE[[i]] = MAE
  df_pred_obs = cbind(predictions,score_df_all_carbo_cor_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  colnames(df_pred_obs)[2] = "obs"
  carbo_cor_plot[[i]] = ggplot(df_pred_obs, aes(x = obs,y = predictions)) + geom_point() + geom_abline(intercept = 0, slope = 1)}
```
  
```{r models doc20 cor }
xenome_human_mouse_RSEM_cor_doc$xeno = rownames(xenome_human_mouse_RSEM_cor_doc)
score_df_all_doc_cor = merge(xenome_human_mouse_RSEM_cor_doc,doc20_quant_score, by = "xeno")
rownames(score_df_all_doc_cor) = score_df_all_doc_cor$xeno
score_df_all_doc_cor$xeno = NULL
colnames(score_df_all_doc_cor)[21] = "response"

score_df_all_doc_cor_qual = merge(xenome_human_mouse_RSEM_cor_doc,doc20_qual_score, by = "xeno")
rownames(score_df_all_doc_cor_qual) = score_df_all_doc_cor_qual$xeno
score_df_all_doc_cor_qual$xeno = NULL
colnames(score_df_all_doc_cor_qual)[21] = "response"

split_10_doc_cor = read.table("input/split_10_doc20_quant.txt")

doc_cor_glm = list()
doc_cor_glm_pred = list()
doc_cor_glm_RMSE = list()
doc_cor_MAE = list()
doc_cor_plot = list()

for (i in 1:10){
  split = split_10_doc_cor[,i]
  score_df_all_doc_cor_train = score_df_all_doc_cor[split,]
  score_df_all_doc_cor_test = score_df_all_doc_cor[-split,]
  glm_model = glm(response ~ ., data = score_df_all_doc_cor_train,maxit = 500)
  doc_cor_glm[[i]] = glm
  predictions <- glm_model %>% predict(score_df_all_doc_cor_test, type = "response")
  df_pred_obs = cbind(predictions,score_df_all_doc_cor_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  doc_cor_glm_pred[[i]] = df_pred_obs
  RMSE = RMSE(predictions,score_df_all_doc_cor_test$response)
  doc_cor_glm_RMSE[[i]] = RMSE
  print(RMSE)
  MAE = MAE(predictions,score_df_all_doc_cor_test$response)
  doc_cor_MAE[[i]] = MAE
  df_pred_obs = cbind(predictions,score_df_all_doc_cor_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  colnames(df_pred_obs)[2] = "obs"
   doc_cor_plot[[i]] = ggplot(df_pred_obs, aes(x = obs,y = predictions)) + geom_point()+ geom_abline(intercept = 0, slope = 1)}
```


#Recursive feature selection

```{r feature selection - rfe carbo }
control_rfe <- rfeControl(functions=rfFuncs, method="cv", number=5,saveDetails = TRUE)
xenome_human_RSEM_all  = xenome_human_RSEM

xenome_human_RSEM_all$xeno = rownames(xenome_human_RSEM_all)
xenome_human_RSEM_all_carbo = merge(xenome_human_RSEM_all,carbo_quant_score, by = "xeno")
xenome_human_RSEM_all_carbo$xeno = NULL
colnames(xenome_human_RSEM_all_carbo)[25385] = "response"
rfe_carbo_ep <- rfe(xenome_human_RSEM_all_carbo[,1:25384], xenome_human_RSEM_all_carbo[,25385], sizes=(10), rfeControl=control_rfe)
predictors(rfe_carbo_ep)
rfe_carbo_genes_ep = predictors(rfe_carbo_ep)
#SV2A is also in this list


#Pick stromal genes with human homologs
xenome_mouse_RSEM_all = xenome_mouse_RSEM
xenome_mouse_RSEM_all_hh = data.frame(t(xenome_mouse_RSEM_all))
xenome_mouse_RSEM_all_hh$genes = rownames(xenome_mouse_RSEM_all_hh)
stromal_hh  = intersect(xenome_mouse_RSEM_all_hh$genes,mouse_to_human_df$genes)
xenome_mouse_RSEM_all_hh = xenome_mouse_RSEM_all_hh[stromal_hh,]
xenome_mouse_RSEM_all_hh$genes = NULL
xenome_mouse_RSEM_all_hh = data.frame(t(xenome_mouse_RSEM_all_hh))
xenome_mouse_RSEM_all_hh$xeno = rownames(xenome_mouse_RSEM_all_hh)
xenome_mouse_RSEM_all_carbo = merge(xenome_mouse_RSEM_all_hh,carbo_quant_score, by = "xeno")

xenome_mouse_RSEM_all_carbo$xeno = NULL
colnames(xenome_mouse_RSEM_all_carbo)[16763] = "response"
rfe_carbo_stromal <- rfe(xenome_mouse_RSEM_all_carbo[,1:16762], xenome_mouse_RSEM_all_carbo[,16763], sizes= c(10), rfeControl=control_rfe)
predictors(rfe_carbo_stromal)
rfe_carbo_genes_stromal = rfe_carbo_stromal$optVariables[1:10]
rfe_carbo_genes_stromal_human = mouse_to_human_df[mouse_to_human_df$genes %in% rfe_carbo_genes_stromal,]
rfe_carbo_genes_stromal_human = as.character(rfe_carbo_genes_stromal_human$HGNC.symbol)


rfe_genes_human_carbo = c(rfe_carbo_genes_ep,rfe_carbo_genes_stromal_human)
rfe_genes_mouse_carbo = c(as.character(rfe_carbo_genes_ep_mouse$genes),rfe_carbo_genes_stromal)
```

```{r feature selection - rfe doc }
xenome_human_RSEM_all$xeno = rownames(xenome_human_RSEM_all)
xenome_human_RSEM_all_doc = merge(xenome_human_RSEM_all,doc20_quant_score, by = "xeno")
xenome_human_RSEM_all_doc$xeno = NULL
colnames(xenome_human_RSEM_all_doc)[25385] = "response"
rfe_doc_ep <- rfe(xenome_human_RSEM_all_doc[,1:25384], xenome_human_RSEM_all_doc[,25385], sizes=(10), rfeControl=control_rfe)
predictors(rfe_doc_ep)
rfe_doc_genes_ep = predictors(rfe_doc_ep)

#Pick stromal genes with human homologs
xenome_mouse_RSEM_all_doc = merge(xenome_mouse_RSEM_all_hh,doc20_quant_score, by = "xeno")
xenome_mouse_RSEM_all_doc$xeno = NULL
colnames(xenome_mouse_RSEM_all_doc)[16763] = "response"
rfe_doc_stromal <- rfe(xenome_mouse_RSEM_all_doc[,1:16762], xenome_mouse_RSEM_all_doc[,16763], sizes= c(10), rfeControl=control_rfe)
predictors(rfe_doc_stromal)
rfe_doc_genes_stromal = rfe_doc_stromal$optVariables[1:10]
rfe_doc_genes_stromal_human = mouse_to_human_df[mouse_to_human_df$genes %in% rfe_doc_genes_stromal,]
rfe_doc_genes_stromal_human = as.character(rfe_doc_genes_stromal_human$HGNC.symbol)


rfe_genes_human_doc = c(rfe_doc_genes_ep,rfe_doc_genes_stromal_human)
rfe_doc_genes_ep_mouse = mouse_to_human_df[mouse_to_human_df$HGNC.symbol %in%  rfe_doc_genes_ep,]
rfe_genes_mouse_doc = c(as.character(rfe_doc_genes_ep_mouse$genes),rfe_doc_genes_stromal)
```

```{r p-bulk rfe carbo}
xenome_human_RSEM_rfe = xenome_human_RSEM[,colnames(xenome_human_RSEM) %in% rfe_genes_human_carbo]
xenome_human_RSEM_rfe$pdx = rownames(xenome_human_RSEM_rfe)
xenome_human_RSEM_rfe_melt = melt(xenome_human_RSEM_rfe)

xenome_mouse_RSEM_rfe = xenome_mouse_RSEM[,colnames(xenome_mouse_RSEM) %in% rfe_genes_mouse_carbo]
#Change colnames to human homologs 
colnames_rfe_mouse = mouse_to_human_df[mouse_to_human_df$genes %in% rfe_genes_mouse_carbo,]
colnames_rfe_mouse_human = as.character(colnames_rfe_mouse$HGNC.symbol)
colnames_rfe_mouse$genes = as.character(colnames_rfe_mouse$genes)
xenome_mouse_RSEM_rfe = xenome_mouse_RSEM[,colnames(xenome_mouse_RSEM) %in% colnames_rfe_mouse$genes]
colnames(xenome_mouse_RSEM_rfe) = colnames_rfe_mouse$HGNC.symbol


xenome_mouse_RSEM_rfe$pdx = rownames(xenome_mouse_RSEM_rfe)
xenome_mouse_RSEM_rfe_melt = melt(xenome_mouse_RSEM_rfe)
xenome_human_mouse_RSEM_rfe_melt = rbind(xenome_human_RSEM_rfe_melt,xenome_mouse_RSEM_rfe_melt)
xenome_human_mouse_RSEM_rfe_melt = aggregate(. ~pdx+variable, data=xenome_human_mouse_RSEM_rfe_melt, sum, na.rm=TRUE)
xenome_human_mouse_RSEM_rfe = recast(xenome_human_mouse_RSEM_rfe_melt, pdx ~ variable)
rownames(xenome_human_mouse_RSEM_rfe) = xenome_human_mouse_RSEM_rfe$pdx
xenome_human_mouse_RSEM_rfe$pdx = NULL 
```

```{r p-bulk rfe  doc}
xenome_human_RSEM_rfe_doc = xenome_human_RSEM[,colnames(xenome_human_RSEM) %in% rfe_genes_human_doc]
xenome_human_RSEM_rfe_doc$pdx = rownames(xenome_human_RSEM_rfe_doc)
xenome_human_RSEM_rfe_doc_melt = melt(xenome_human_RSEM_rfe_doc)

xenome_mouse_RSEM_rfe_doc = xenome_mouse_RSEM[,colnames(xenome_mouse_RSEM) %in% rfe_genes_mouse_doc]
#Change colnames to human homologs 
colnames_rfe_doc_mouse = mouse_to_human_df[mouse_to_human_df$genes %in% rfe_genes_mouse_doc,]
colnames_rfe_doc_mouse_human = as.character(colnames_rfe_doc_mouse$HGNC.symbol)
colnames_rfe_doc_mouse$genes = as.character(colnames_rfe_doc_mouse$genes)
doc_xenome_mouse = intersect(colnames(xenome_mouse_RSEM),colnames_rfe_doc_mouse$genes)

colnames_rfe_doc_mouse  = colnames_rfe_doc_mouse[colnames_rfe_doc_mouse$genes %in% doc_xenome_mouse, ]
xenome_mouse_RSEM_rfe_doc = xenome_mouse_RSEM[,colnames_rfe_doc_mouse$genes]
colnames(xenome_mouse_RSEM_rfe_doc) = colnames_rfe_doc_mouse$HGNC.symbol


xenome_mouse_RSEM_rfe_doc$pdx = rownames(xenome_mouse_RSEM_rfe_doc)
xenome_mouse_RSEM_rfe_doc_melt = melt(xenome_mouse_RSEM_rfe_doc)
xenome_human_mouse_RSEM_rfe_doc_melt = rbind(xenome_human_RSEM_rfe_doc_melt,xenome_mouse_RSEM_rfe_doc_melt)
xenome_human_mouse_RSEM_rfe_doc_melt = aggregate(. ~pdx+variable, data=xenome_human_mouse_RSEM_rfe_doc_melt, sum, na.rm=TRUE)
xenome_human_mouse_RSEM_rfe_doc = recast(xenome_human_mouse_RSEM_rfe_doc_melt, pdx ~ variable)
rownames(xenome_human_mouse_RSEM_rfe_doc) = xenome_human_mouse_RSEM_rfe_doc$pdx
xenome_human_mouse_RSEM_rfe_doc$pdx = NULL 
```


```{r  models carbo rfe}
#Can we predict these scores?
xenome_human_mouse_RSEM_rfe$xeno = rownames(xenome_human_mouse_RSEM_rfe)
score_df_all_carbo_rfe = merge(xenome_human_mouse_RSEM_rfe,carbo_quant_score, by = "xeno")
rownames(score_df_all_carbo_rfe) = score_df_all_carbo_rfe$xeno
score_df_all_carbo_rfe$xeno = NULL
colnames(score_df_all_carbo_rfe)[21] = "response"

split_10_carbo_rfe = read.table("input/split_10_carbo_quant.txt")
carbo_rfe_glm = list()
carbo_rfe_glm_pred = list()
carbo_rfe_glm_RMSE = list()
carbo_rfe_MAE = list()
carbo_rfe_plot = list()

for (i in 1:10){
  split = split_10_carbo_rfe[,i]
  score_df_all_carbo_rfe_train = score_df_all_carbo_rfe[split,]
  score_df_all_carbo_rfe_test = score_df_all_carbo_rfe[-split,]
  glm_model = glm(response ~ ., data = score_df_all_carbo_rfe_train,maxit = 500)
  carbo_rfe_glm[[i]] = glm
  predictions <- glm_model %>% predict(score_df_all_carbo_rfe_test, type = "response")
  df_pred_obs = cbind(predictions,score_df_all_carbo_rfe_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  carbo_rfe_glm_pred[[i]] = df_pred_obs
  RMSE = RMSE(predictions,score_df_all_carbo_rfe_test$response)
  carbo_rfe_glm_RMSE[[i]] = RMSE
  print(RMSE)
  MAE = MAE(predictions,score_df_all_carbo_rfe_test$response)
  carbo_rfe_MAE[[i]] = MAE
  df_pred_obs = cbind(predictions,score_df_all_carbo_rfe_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  colnames(df_pred_obs)[2] = "obs"
  carbo_rfe_plot[[i]] = ggplot(df_pred_obs, aes(x = obs,y = predictions)) + geom_point() + geom_abline(intercept = 0, slope = 1)}
```
  
```{r models doc20 rfe }

xenome_human_mouse_RSEM_rfe_doc$xeno = rownames(xenome_human_mouse_RSEM_rfe_doc)
score_df_all_doc_rfe = merge(xenome_human_mouse_RSEM_rfe_doc,doc20_quant_score, by = "xeno")
rownames(score_df_all_doc_rfe) = score_df_all_doc_rfe$xeno
score_df_all_doc_rfe$xeno = NULL
colnames(score_df_all_doc_rfe)[20] = "response"

split_10_doc_rfe = read.table("input/split_10_doc20_quant.txt")

doc_rfe_glm = list()
doc_rfe_glm_pred = list()
doc_rfe_glm_RMSE = list()
doc_rfe_MAE = list()
doc_rfe_plot = list()

for (i in 1:10){
  split = split_10_doc_rfe[,i]
  score_df_all_doc_rfe_train = score_df_all_doc_rfe[split,]
  score_df_all_doc_rfe_test = score_df_all_doc_rfe[-split,]
  glm_model = glm(response ~ ., data = score_df_all_doc_rfe_train,maxit = 500)
  doc_rfe_glm[[i]] = glm
  predictions <- glm_model %>% predict(score_df_all_doc_rfe_test, type = "response")
  df_pred_obs = cbind(predictions,score_df_all_doc_rfe_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  doc_rfe_glm_pred[[i]] = df_pred_obs
  RMSE = RMSE(predictions,score_df_all_doc_rfe_test$response)
  doc_rfe_glm_RMSE[[i]] = RMSE
  print(RMSE)
  MAE = MAE(predictions,score_df_all_doc_rfe_test$response)
  doc_rfe_MAE[[i]] = MAE
  df_pred_obs = cbind(predictions,score_df_all_doc_rfe_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  colnames(df_pred_obs)[2] = "obs"
   doc_rfe_plot[[i]] = ggplot(df_pred_obs, aes(x = obs,y = predictions)) + geom_point()+ geom_abline(intercept = 0, slope = 1)}
```

#Boruta

```{r feature selection - boruta carbo }
#Boruto does not work with 20k + genes
xenome_human_RSEM_5000 = read.table("input/xenome_human_RSEM_log2_5000_var.txt")
xenome_human_RSEM_5000$xeno = rownames(xenome_human_RSEM_5000)
xenome_human_RSEM_all_carbo_5000 = merge(xenome_human_RSEM_5000,carbo_quant_score, by = "xeno")
rownames(xenome_human_RSEM_all_carbo_5000) = xenome_human_RSEM_all_carbo_5000$xeno
xenome_human_RSEM_all_carbo_5000$xeno = NULL
colnames(xenome_human_RSEM_all_carbo_5000)[5001] = "response"

boruta_carbo_ep = Boruta::Boruta(response ~ ., data = xenome_human_RSEM_all_carbo_5000)
boruta_carbo_ep_imp = data.frame(boruta_carbo_ep$finalDecision)
boruta_carbo_ep_imp = boruta_carbo_ep_imp[boruta_carbo_ep_imp$boruta_carbo_ep.finalDecision %in% c("Tentative","Confirmed"),,drop = FALSE ]
boruta_carbo_ep_imp = rownames(boruta_carbo_ep_imp)


xenome_mouse_RSEM_5000 = read.table("input/xenome_mouse_RSEM_log2_5000_var.txt")
xenome_mouse_RSEM_5000$xeno = rownames(xenome_mouse_RSEM_5000)
xenome_mouse_RSEM_all_carbo_5000 = merge(xenome_mouse_RSEM_5000,carbo_quant_score, by = "xeno")
rownames(xenome_mouse_RSEM_all_carbo_5000) = xenome_mouse_RSEM_all_carbo_5000$xeno
xenome_mouse_RSEM_all_carbo_5000$xeno = NULL
colnames(xenome_mouse_RSEM_all_carbo_5000)[5001] = "response"

boruta_carbo_stromal = Boruta::Boruta(response ~ ., data = xenome_mouse_RSEM_all_carbo_5000)
boruta_carbo_stromal_imp = data.frame(boruta_carbo_stromal$finalDecision)
boruta_carbo_stromal_imp = boruta_carbo_stromal_imp[boruta_carbo_stromal_imp$boruta_carbo_stromal.finalDecision %in% c("Tentative","Confirmed"),,drop = FALSE ]
boruta_carbo_stromal_imp = rownames(boruta_carbo_stromal_imp)
boruta_carbo_stromal_imp = boruta_carbo_stromal_imp[-1]

```

```{r feature selection - boruta doc }
xenome_human_RSEM_5000$xeno = rownames(xenome_human_RSEM_5000)
xenome_human_RSEM_all_doc_5000 = merge(xenome_human_RSEM_5000,doc20_quant_score, by = "xeno")
rownames(xenome_human_RSEM_all_doc_5000) = xenome_human_RSEM_all_doc_5000$xeno
xenome_human_RSEM_all_doc_5000$xeno = NULL
colnames(xenome_human_RSEM_all_doc_5000)[5001] = "response"

boruta_doc_ep = Boruta::Boruta(response ~ ., data = xenome_human_RSEM_all_doc_5000)
boruta_doc_ep_imp = data.frame(boruta_doc_ep$finalDecision)
boruta_doc_ep_imp = boruta_doc_ep_imp[boruta_doc_ep_imp$boruta_doc_ep.finalDecision %in% c("Tentative","Confirmed"),,drop = FALSE ]
boruta_doc_ep_imp = rownames(boruta_doc_ep_imp)

xenome_mouse_RSEM_5000$xeno = rownames(xenome_mouse_RSEM_5000)
xenome_mouse_RSEM_all_doc_5000 = merge(xenome_mouse_RSEM_5000,doc20_quant_score, by = "xeno")
rownames(xenome_mouse_RSEM_all_doc_5000) = xenome_mouse_RSEM_all_doc_5000$xeno
xenome_mouse_RSEM_all_doc_5000$xeno = NULL
colnames(xenome_mouse_RSEM_all_doc_5000)[5001] = "response"

boruta_doc_stromal = Boruta::Boruta(response ~ ., data = xenome_mouse_RSEM_all_doc_5000)
boruta_doc_stromal_imp = data.frame(boruta_doc_stromal$finalDecision)
boruta_doc_stromal_imp = boruta_doc_stromal_imp[boruta_doc_stromal_imp$boruta_doc_stromal.finalDecision %in% c("Tentative","Confirmed"),,drop = FALSE ]
boruta_doc_stromal_imp = rownames(boruta_doc_stromal_imp)
```

```{r p-bulk boruta carbo}
xenome_human_RSEM_bortua_carbo = xenome_human_RSEM[,colnames(xenome_human_RSEM) %in% boruta_carbo_ep_imp]
xenome_human_RSEM_bortua_carbo$pdx = rownames(xenome_human_RSEM_bortua_carbo)
xenome_human_RSEM_bortua_carbo_melt = melt(xenome_human_RSEM_bortua_carbo)

xenome_mouse_RSEM_bortua_carbo = xenome_mouse_RSEM[,colnames(xenome_mouse_RSEM) %in% boruta_carbo_stromal_imp]
#Change colnames to human homologs 
colnames_bortua_carbo_mouse = mouse_to_human_df[mouse_to_human_df$genes %in% boruta_carbo_stromal_imp,]
colnames_bortua_carbo_mouse_human = as.character(colnames_bortua_carbo_mouse$HGNC.symbol)
colnames_bortua_carbo_mouse$genes = as.character(colnames_bortua_carbo_mouse$genes)
xenome_mouse_RSEM_bortua_carbo = xenome_mouse_RSEM[,colnames_bortua_carbo_mouse$genes]
colnames(xenome_mouse_RSEM_bortua_carbo) = colnames_bortua_carbo_mouse$HGNC.symbol


xenome_mouse_RSEM_bortua_carbo$pdx = rownames(xenome_mouse_RSEM_bortua_carbo)
xenome_mouse_RSEM_bortua_carbo_melt = melt(xenome_mouse_RSEM_bortua_carbo)
xenome_human_mouse_RSEM_bortua_carbo_melt = rbind(xenome_human_RSEM_bortua_carbo_melt,xenome_mouse_RSEM_bortua_carbo_melt)
xenome_human_mouse_RSEM_bortua_carbo_melt = aggregate(. ~pdx+variable, data=xenome_human_mouse_RSEM_bortua_carbo_melt, sum, na.rm=TRUE)
xenome_human_mouse_RSEM_bortua_carbo = recast(xenome_human_mouse_RSEM_bortua_carbo_melt, pdx ~ variable)
rownames(xenome_human_mouse_RSEM_bortua_carbo) = xenome_human_mouse_RSEM_bortua_carbo$pdx
xenome_human_mouse_RSEM_bortua_carbo$pdx = NULL 
```


```{r p-bulk boruta doc}
xenome_human_RSEM_bortua_doc = xenome_human_RSEM[,colnames(xenome_human_RSEM) %in% boruta_doc_ep_imp]
xenome_human_RSEM_bortua_doc$pdx = rownames(xenome_human_RSEM_bortua_doc)
xenome_human_RSEM_bortua_doc_melt = melt(xenome_human_RSEM_bortua_doc)

xenome_mouse_RSEM_bortua_doc = xenome_mouse_RSEM[,colnames(xenome_mouse_RSEM) %in% boruta_doc_stromal_imp]
#Change colnames to human homologs 
colnames_bortua_doc_mouse = mouse_to_human_df[mouse_to_human_df$genes %in% boruta_doc_stromal_imp,]
colnames_bortua_doc_mouse_human = as.character(colnames_bortua_doc_mouse$HGNC.symbol)
colnames_bortua_doc_mouse$genes = as.character(colnames_bortua_doc_mouse$genes)
xenome_mouse_RSEM_bortua_doc = xenome_mouse_RSEM[,colnames_bortua_doc_mouse$genes]
colnames(xenome_mouse_RSEM_bortua_doc) = colnames_bortua_doc_mouse$HGNC.symbol


xenome_mouse_RSEM_bortua_doc$pdx = rownames(xenome_mouse_RSEM_bortua_doc)
xenome_mouse_RSEM_bortua_doc_melt = melt(xenome_mouse_RSEM_bortua_doc)
xenome_human_mouse_RSEM_bortua_doc_melt = rbind(xenome_human_RSEM_bortua_doc_melt,xenome_mouse_RSEM_bortua_doc_melt)
xenome_human_mouse_RSEM_bortua_doc_melt = aggregate(. ~pdx+variable, data=xenome_human_mouse_RSEM_bortua_doc_melt, sum, na.rm=TRUE)
xenome_human_mouse_RSEM_bortua_doc = recast(xenome_human_mouse_RSEM_bortua_doc_melt, pdx ~ variable)
rownames(xenome_human_mouse_RSEM_bortua_doc) = xenome_human_mouse_RSEM_bortua_doc$pdx
xenome_human_mouse_RSEM_bortua_doc$pdx = NULL 
```


```{r  models carbo boruta}
#Can we predict these scores?
xenome_human_mouse_RSEM_bortua_carbo$xeno = rownames(xenome_human_mouse_RSEM_bortua_carbo)
score_df_all_carbo_boruta = merge(xenome_human_mouse_RSEM_bortua_carbo,carbo_quant_score, by = "xeno")
rownames(score_df_all_carbo_boruta) = score_df_all_carbo_boruta$xeno
score_df_all_carbo_boruta$xeno = NULL
colnames(score_df_all_carbo_boruta)[23] = "response"

split_10_carbo_boruta = read.table("input/split_10_carbo_quant.txt")
carbo_boruta_glm = list()
carbo_boruta_glm_pred = list()
carbo_boruta_glm_RMSE = list()
carbo_boruta_MAE = list()
carbo_boruta_plot = list()

for (i in 1:10){
  split = split_10_carbo_boruta[,i]
  score_df_all_carbo_boruta_train = score_df_all_carbo_boruta[split,]
  score_df_all_carbo_boruta_test = score_df_all_carbo_boruta[-split,]
  glm_model = glm(response ~ ., data = score_df_all_carbo_boruta_train,maxit = 500)
  carbo_boruta_glm[[i]] = glm
  predictions <- glm_model %>% predict(score_df_all_carbo_boruta_test, type = "response")
  df_pred_obs = cbind(predictions,score_df_all_carbo_boruta_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  carbo_boruta_glm_pred[[i]] = df_pred_obs
  RMSE = RMSE(predictions,score_df_all_carbo_boruta_test$response)
  carbo_boruta_glm_RMSE[[i]] = RMSE
  print(RMSE)
  MAE = MAE(predictions,score_df_all_carbo_boruta_test$response)
  carbo_boruta_MAE[[i]] = MAE
  df_pred_obs = cbind(predictions,score_df_all_carbo_boruta_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  colnames(df_pred_obs)[2] = "obs"
  carbo_boruta_plot[[i]] = ggplot(df_pred_obs, aes(x = obs,y = predictions)) + geom_point() + geom_abline(intercept = 0, slope = 1)}
```
  
```{r models doc20 boruta }

xenome_human_mouse_RSEM_bortua_doc$xeno = rownames(xenome_human_mouse_RSEM_bortua_doc)
score_df_all_doc_boruta = merge(xenome_human_mouse_RSEM_bortua_doc,doc20_quant_score, by = "xeno")
rownames(score_df_all_doc_boruta) = score_df_all_doc_boruta$xeno
score_df_all_doc_boruta$xeno = NULL
colnames(score_df_all_doc_boruta)[29] = "response"

split_10_doc_boruta = read.table("input/split_10_doc20_quant.txt")
doc_boruta_glm = list()
doc_boruta_glm_pred = list()
doc_boruta_glm_RMSE = list()
doc_boruta_MAE = list()
doc_boruta_plot = list()

for (i in 1:10){
  split = split_10_doc_boruta[,i]
  score_df_all_doc_boruta_train = score_df_all_doc_boruta[split,]
  score_df_all_doc_boruta_test = score_df_all_doc_boruta[-split,]
  glm_model = glm(response ~ ., data = score_df_all_doc_boruta_train,maxit = 500)
  doc_boruta_glm[[i]] = glm
  predictions <- glm_model %>% predict(score_df_all_doc_boruta_test, type = "response")
  df_pred_obs = cbind(predictions,score_df_all_doc_boruta_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  doc_boruta_glm_pred[[i]] = df_pred_obs
  RMSE = RMSE(predictions,score_df_all_doc_boruta_test$response)
  doc_boruta_glm_RMSE[[i]] = RMSE
  print(RMSE)
  MAE = MAE(predictions,score_df_all_doc_boruta_test$response)
  doc_boruta_MAE[[i]] = MAE
  df_pred_obs = cbind(predictions,score_df_all_doc_boruta_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  colnames(df_pred_obs)[2] = "obs"
   doc_boruta_plot[[i]] = ggplot(df_pred_obs, aes(x = obs,y = predictions)) + geom_point()+ geom_abline(intercept = 0, slope = 1)}
```

#Sklean
```{python feature selection carbo f regression}
from pandas import read_csv
import pandas as pd 
from numpy import set_printoptions
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif
from sklearn.feature_selection import f_regression
from sklearn.feature_selection import mutual_info_regression
# load data

#Carbo_ep
dataframe = r.xenome_human_RSEM_all_carbo_5000
array = dataframe.values
# feature extraction
X = array[:,0:5000]
Y = array[:,5000]

#f test
test_f = SelectKBest(score_func=f_regression, k=10)
fit_f = test_f.fit(X, Y)
# summarize scores
carbo_ep_cols_f = test_f.get_support(indices=True)
df_new_f = X[:,carbo_ep_cols_f]
# summarize selected features
print(carbo_ep_cols_f)

#mutal_info regression
test_m = SelectKBest(score_func=mutual_info_regression, k=10)
fit_m = test_m.fit(X, Y)
# summarize scores
carbo_ep_cols_m = test_m.get_support(indices=True)
df_new_m = X[:,carbo_ep_cols_m]
# summarize selected meatures
print(carbo_ep_cols_m)



#Carbo_stromal
dataframe_mouse = r.xenome_mouse_RSEM_all_carbo_5000
array_mouse = dataframe_mouse.values
# feature extraction
X_mouse = array[:,0:5000]
Y_mouse = array[:,5000]

#f test
test_f_mouse = SelectKBest(score_func=f_regression, k=10)
fit_f_mouse = test_f_mouse.fit(X_mouse,Y_mouse)
# summarize scores
carbo_stromal_cols_f = test_f_mouse.get_support(indices=True)
df_new_f_mouse = X_mouse[:,carbo_stromal_cols_f]
# summarize selected features
print(carbo_stromal_cols_f)

#mutal_info regression
test_m_mouse = SelectKBest(score_func=mutual_info_regression, k=10)
fit_m_mouse = test_m_mouse.fit(X_mouse, Y_mouse)
# summarize scores
carbo_stromal_cols_m = test_m_mouse.get_support(indices=True)
df_new_m_mouse = X_mouse[:,carbo_stromal_cols_m]
# summarize selected meatures
print(carbo_stromal_cols_m)


#doc_ep
dataframe = r.xenome_human_RSEM_all_doc_5000
array = dataframe.values
# feature extraction
X = array[:,0:5000]
Y = array[:,5000]

#f test
test_f = SelectKBest(score_func=f_regression, k=10)
fit_f = test_f.fit(X, Y)
# summarize scores
doc_ep_cols_f = test_f.get_support(indices=True)
df_new_f = X[:,doc_ep_cols_f]
# summarize selected features
print(doc_ep_cols_f)

#mutal_info regression
test_m = SelectKBest(score_func=mutual_info_regression, k=10)
fit_m = test_m.fit(X, Y)
# summarize scores
doc_ep_cols_m = test_m.get_support(indices=True)
df_new_m = X[:,doc_ep_cols_m]
# summarize selected meatures
print(doc_ep_cols_m)



#doc_stromal
dataframe_mouse = r.xenome_mouse_RSEM_all_doc_5000
array_mouse = dataframe_mouse.values
# feature extraction
X_mouse = array[:,0:5000]
Y_mouse = array[:,5000]

#f test
test_f_mouse = SelectKBest(score_func=f_regression, k=10)
fit_f_mouse = test_f_mouse.fit(X_mouse,Y_mouse)
# summarize scores
doc_stromal_cols_f = test_f_mouse.get_support(indices=True)
df_new_f_mouse = X_mouse[:,doc_stromal_cols_f]
# summarize selected features
print(doc_stromal_cols_f)

#mutal_info regression
test_m_mouse = SelectKBest(score_func=mutual_info_regression, k=10)
fit_m_mouse = test_m_mouse.fit(X_mouse, Y_mouse)
# summarize scores
doc_stromal_cols_m = test_m_mouse.get_support(indices=True)
df_new_m_mouse = X_mouse[:,doc_stromal_cols_m]
# summarize selected meatures
print(doc_stromal_cols_m)

```


```{r sklearn}
sklearn_f_carbo_ep_imp = colnames(xenome_human_RSEM_all_carbo_5000)[py$carbo_ep_cols_f + 1]
sklearn_m_carbo_ep_imp = colnames(xenome_human_RSEM_all_carbo_5000)[py$carbo_ep_cols_m + 1]
sklearn_f_carbo_stromal_imp = colnames(xenome_mouse_RSEM_all_carbo_5000)[py$carbo_stromal_cols_f + 1]
sklearn_f_carbo_stromal_imp = sklearn_f_carbo_stromal_imp[-8]
sklearn_m_carbo_stromal_imp = colnames(xenome_mouse_RSEM_all_carbo_5000)[py$carbo_stromal_cols_m + 1]
sklearn_f_carbo_stromal_imp = sklearn_f_carbo_stromal_imp[-7]


sklearn_f_doc_ep_imp = colnames(xenome_human_RSEM_all_doc_5000)[py$doc_ep_cols_f + 1]
sklearn_m_doc_ep_imp = colnames(xenome_human_RSEM_all_doc_5000)[py$doc_ep_cols_m + 1]
sklearn_f_doc_stromal_imp = colnames(xenome_mouse_RSEM_all_doc_5000)[py$doc_stromal_cols_f + 1]
sklearn_f_doc_stromal_imp = sklearn_f_doc_stromal_imp[-3]
sklearn_f_doc_stromal_imp = sklearn_f_doc_stromal_imp[-7]


sklearn_m_doc_stromal_imp = colnames(xenome_mouse_RSEM_all_doc_5000)[py$doc_stromal_cols_m + 1]
sklearn_m_doc_stromal_imp = sklearn_m_doc_stromal_imp[-4]
sklearn_m_doc_stromal_imp = sklearn_m_doc_stromal_imp[-3]

```


```{r p-bulk sklearn_f carbo}
xenome_human_RSEM_sklearn_f_carbo = xenome_human_RSEM[,colnames(xenome_human_RSEM) %in% sklearn_f_carbo_ep_imp]
xenome_human_RSEM_sklearn_f_carbo$pdx = rownames(xenome_human_RSEM_sklearn_f_carbo)
xenome_human_RSEM_sklearn_f_carbo_melt = melt(xenome_human_RSEM_sklearn_f_carbo)

xenome_mouse_RSEM_sklearn_f_carbo = xenome_mouse_RSEM[,colnames(xenome_mouse_RSEM) %in% sklearn_f_carbo_stromal_imp]
#Change colnames to human homologs 
colnames_sklearn_f_carbo_mouse = mouse_to_human_df[mouse_to_human_df$genes %in% sklearn_f_carbo_stromal_imp,]
colnames_sklearn_f_carbo_mouse_human = as.character(colnames_sklearn_f_carbo_mouse$HGNC.symbol)
colnames_sklearn_f_carbo_mouse$genes = as.character(colnames_sklearn_f_carbo_mouse$genes)
xenome_mouse_RSEM_sklearn_f_carbo = xenome_mouse_RSEM[,colnames_sklearn_f_carbo_mouse$genes]
colnames(xenome_mouse_RSEM_sklearn_f_carbo) = colnames_sklearn_f_carbo_mouse$HGNC.symbol


xenome_mouse_RSEM_sklearn_f_carbo$pdx = rownames(xenome_mouse_RSEM_sklearn_f_carbo)
xenome_mouse_RSEM_sklearn_f_carbo_melt = melt(xenome_mouse_RSEM_sklearn_f_carbo)
xenome_human_mouse_RSEM_sklearn_f_carbo_melt = rbind(xenome_human_RSEM_sklearn_f_carbo_melt,xenome_mouse_RSEM_sklearn_f_carbo_melt)
xenome_human_mouse_RSEM_sklearn_f_carbo_melt = aggregate(. ~pdx+variable, data=xenome_human_mouse_RSEM_sklearn_f_carbo_melt, sum, na.rm=TRUE)
xenome_human_mouse_RSEM_sklearn_f_carbo = recast(xenome_human_mouse_RSEM_sklearn_f_carbo_melt, pdx ~ variable)
rownames(xenome_human_mouse_RSEM_sklearn_f_carbo) = xenome_human_mouse_RSEM_sklearn_f_carbo$pdx
xenome_human_mouse_RSEM_sklearn_f_carbo$pdx = NULL 
```

```{r p-bulk sklearn_m carbo}
xenome_human_RSEM_sklearn_m_carbo = xenome_human_RSEM[,colnames(xenome_human_RSEM) %in% sklearn_m_carbo_ep_imp]
xenome_human_RSEM_sklearn_m_carbo$pdx = rownames(xenome_human_RSEM_sklearn_m_carbo)
xenome_human_RSEM_sklearn_m_carbo_melt = melt(xenome_human_RSEM_sklearn_m_carbo)

xenome_mouse_RSEM_sklearn_m_carbo = xenome_mouse_RSEM[,colnames(xenome_mouse_RSEM) %in% sklearn_m_carbo_stromal_imp]
#Change colnames to human homologs 
colnames_sklearn_m_carbo_mouse = mouse_to_human_df[mouse_to_human_df$genes %in% sklearn_m_carbo_stromal_imp,]
colnames_sklearn_m_carbo_mouse_human = as.character(colnames_sklearn_m_carbo_mouse$HGNC.symbol)
colnames_sklearn_m_carbo_mouse$genes = as.character(colnames_sklearn_m_carbo_mouse$genes)
xenome_mouse_RSEM_sklearn_m_carbo = xenome_mouse_RSEM[,colnames_sklearn_m_carbo_mouse$genes]
colnames(xenome_mouse_RSEM_sklearn_m_carbo) = colnames_sklearn_m_carbo_mouse$HGNC.symbol


xenome_mouse_RSEM_sklearn_m_carbo$pdx = rownames(xenome_mouse_RSEM_sklearn_m_carbo)
xenome_mouse_RSEM_sklearn_m_carbo_melt = melt(xenome_mouse_RSEM_sklearn_m_carbo)
xenome_human_mouse_RSEM_sklearn_m_carbo_melt = rbind(xenome_human_RSEM_sklearn_m_carbo_melt,xenome_mouse_RSEM_sklearn_m_carbo_melt)
xenome_human_mouse_RSEM_sklearn_m_carbo_melt = aggregate(. ~pdx+variable, data=xenome_human_mouse_RSEM_sklearn_m_carbo_melt, sum, na.rm=TRUE)
xenome_human_mouse_RSEM_sklearn_m_carbo = recast(xenome_human_mouse_RSEM_sklearn_m_carbo_melt, pdx ~ variable)
rownames(xenome_human_mouse_RSEM_sklearn_m_carbo) = xenome_human_mouse_RSEM_sklearn_m_carbo$pdx
xenome_human_mouse_RSEM_sklearn_m_carbo$pdx = NULL 
```



```{r p-bulk sklearn_f doc}
xenome_human_RSEM_sklearn_f_doc = xenome_human_RSEM[,colnames(xenome_human_RSEM) %in% sklearn_f_doc_ep_imp]
xenome_human_RSEM_sklearn_f_doc$pdx = rownames(xenome_human_RSEM_sklearn_f_doc)
xenome_human_RSEM_sklearn_f_doc_melt = melt(xenome_human_RSEM_sklearn_f_doc)

xenome_mouse_RSEM_sklearn_f_doc = xenome_mouse_RSEM[,colnames(xenome_mouse_RSEM) %in% sklearn_f_doc_stromal_imp]
#Change colnames to human homologs 
colnames_sklearn_f_doc_mouse = mouse_to_human_df[mouse_to_human_df$genes %in% sklearn_f_doc_stromal_imp,]
colnames_sklearn_f_doc_mouse_human = as.character(colnames_sklearn_f_doc_mouse$HGNC.symbol)
colnames_sklearn_f_doc_mouse$genes = as.character(colnames_sklearn_f_doc_mouse$genes)
xenome_mouse_RSEM_sklearn_f_doc = xenome_mouse_RSEM[,colnames_sklearn_f_doc_mouse$genes]
colnames(xenome_mouse_RSEM_sklearn_f_doc) = colnames_sklearn_f_doc_mouse$HGNC.symbol


xenome_mouse_RSEM_sklearn_f_doc$pdx = rownames(xenome_mouse_RSEM_sklearn_f_doc)
xenome_mouse_RSEM_sklearn_f_doc_melt = melt(xenome_mouse_RSEM_sklearn_f_doc)
xenome_human_mouse_RSEM_sklearn_f_doc_melt = rbind(xenome_human_RSEM_sklearn_f_doc_melt,xenome_mouse_RSEM_sklearn_f_doc_melt)
xenome_human_mouse_RSEM_sklearn_f_doc_melt = aggregate(. ~pdx+variable, data=xenome_human_mouse_RSEM_sklearn_f_doc_melt, sum, na.rm=TRUE)
xenome_human_mouse_RSEM_sklearn_f_doc = recast(xenome_human_mouse_RSEM_sklearn_f_doc_melt, pdx ~ variable)
rownames(xenome_human_mouse_RSEM_sklearn_f_doc) = xenome_human_mouse_RSEM_sklearn_f_doc$pdx
xenome_human_mouse_RSEM_sklearn_f_doc$pdx = NULL 
```

```{r p-bulk sklearn_m doc}
xenome_human_RSEM_sklearn_m_doc = xenome_human_RSEM[,colnames(xenome_human_RSEM) %in% sklearn_m_doc_ep_imp]
xenome_human_RSEM_sklearn_m_doc$pdx = rownames(xenome_human_RSEM_sklearn_m_doc)
xenome_human_RSEM_sklearn_m_doc_melt = melt(xenome_human_RSEM_sklearn_m_doc)

xenome_mouse_RSEM_sklearn_m_doc = xenome_mouse_RSEM[,colnames(xenome_mouse_RSEM) %in% sklearn_m_doc_stromal_imp]
#Change colnames to human homologs 
colnames_sklearn_m_doc_mouse = mouse_to_human_df[mouse_to_human_df$genes %in% sklearn_m_doc_stromal_imp,]
colnames_sklearn_m_doc_mouse_human = as.character(colnames_sklearn_m_doc_mouse$HGNC.symbol)
colnames_sklearn_m_doc_mouse$genes = as.character(colnames_sklearn_m_doc_mouse$genes)
xenome_mouse_RSEM_sklearn_m_doc = xenome_mouse_RSEM[,colnames_sklearn_m_doc_mouse$genes]
colnames(xenome_mouse_RSEM_sklearn_m_doc) = colnames_sklearn_m_doc_mouse$HGNC.symbol


xenome_mouse_RSEM_sklearn_m_doc$pdx = rownames(xenome_mouse_RSEM_sklearn_m_doc)
xenome_mouse_RSEM_sklearn_m_doc_melt = melt(xenome_mouse_RSEM_sklearn_m_doc)
xenome_human_mouse_RSEM_sklearn_m_doc_melt = rbind(xenome_human_RSEM_sklearn_m_doc_melt,xenome_mouse_RSEM_sklearn_m_doc_melt)
xenome_human_mouse_RSEM_sklearn_m_doc_melt = aggregate(. ~pdx+variable, data=xenome_human_mouse_RSEM_sklearn_m_doc_melt, sum, na.rm=TRUE)
xenome_human_mouse_RSEM_sklearn_m_doc = recast(xenome_human_mouse_RSEM_sklearn_m_doc_melt, pdx ~ variable)
rownames(xenome_human_mouse_RSEM_sklearn_m_doc) = xenome_human_mouse_RSEM_sklearn_m_doc$pdx
xenome_human_mouse_RSEM_sklearn_m_doc$pdx = NULL 
```



```{r  models carbo sklearn_f}
xenome_human_mouse_RSEM_sklearn_f_carbo$xeno = rownames(xenome_human_mouse_RSEM_sklearn_f_carbo)
score_df_all_carbo_sklearn_f = merge(xenome_human_mouse_RSEM_sklearn_f_carbo,carbo_quant_score, by = "xeno")
rownames(score_df_all_carbo_sklearn_f) = score_df_all_carbo_sklearn_f$xeno
score_df_all_carbo_sklearn_f$xeno = NULL
colnames(score_df_all_carbo_sklearn_f)[19] = "response"

split_10_carbo_sklearn_f = read.table("input/split_10_carbo_quant.txt")
carbo_sklearn_f_glm = list()
carbo_sklearn_f_glm_pred = list()
carbo_sklearn_f_glm_RMSE = list()
carbo_sklearn_f_MAE = list()
carbo_sklearn_f_plot = list()

for (i in 1:10){
  split = split_10_carbo_sklearn_f[,i]
  score_df_all_carbo_sklearn_f_train = score_df_all_carbo_sklearn_f[split,]
  score_df_all_carbo_sklearn_f_test = score_df_all_carbo_sklearn_f[-split,]
  glm_model = glm(response ~ ., data = score_df_all_carbo_sklearn_f_train,maxit = 500)
  carbo_sklearn_f_glm[[i]] = glm
  predictions <- glm_model %>% predict(score_df_all_carbo_sklearn_f_test, type = "response")
  df_pred_obs = cbind(predictions,score_df_all_carbo_sklearn_f_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  carbo_sklearn_f_glm_pred[[i]] = df_pred_obs
  RMSE = RMSE(predictions,score_df_all_carbo_sklearn_f_test$response)
  carbo_sklearn_f_glm_RMSE[[i]] = RMSE
  print(RMSE)
  MAE = MAE(predictions,score_df_all_carbo_sklearn_f_test$response)
  carbo_sklearn_f_MAE[[i]] = MAE
  df_pred_obs = cbind(predictions,score_df_all_carbo_sklearn_f_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  colnames(df_pred_obs)[2] = "obs"
  carbo_sklearn_f_plot[[i]] = ggplot(df_pred_obs, aes(x = obs,y = predictions)) + geom_point() + geom_abline(intercept = 0, slope = 1)}
```
  
  
```{r  models carbo sklearn_m}
xenome_human_mouse_RSEM_sklearn_m_carbo$xeno = rownames(xenome_human_mouse_RSEM_sklearn_m_carbo)
score_df_all_carbo_sklearn_m = merge(xenome_human_mouse_RSEM_sklearn_m_carbo,carbo_quant_score, by = "xeno")
rownames(score_df_all_carbo_sklearn_m) = score_df_all_carbo_sklearn_m$xeno
score_df_all_carbo_sklearn_m$xeno = NULL
colnames(score_df_all_carbo_sklearn_m)[17] = "response"

split_10_carbo_sklearn_m = read.table("input/split_10_carbo_quant.txt")
carbo_sklearn_m_glm = list()
carbo_sklearn_m_glm_pred = list()
carbo_sklearn_m_glm_RMSE = list()
carbo_sklearn_m_MAE = list()
carbo_sklearn_m_plot = list()

for (i in 1:10){
  split = split_10_carbo_sklearn_m[,i]
  score_df_all_carbo_sklearn_m_train = score_df_all_carbo_sklearn_m[split,]
  score_df_all_carbo_sklearn_m_test = score_df_all_carbo_sklearn_m[-split,]
  glm_model = glm(response ~ ., data = score_df_all_carbo_sklearn_m_train,maxit = 500)
  carbo_sklearn_m_glm[[i]] = glm
  predictions <- glm_model %>% predict(score_df_all_carbo_sklearn_m_test, type = "response")
  df_pred_obs = cbind(predictions,score_df_all_carbo_sklearn_m_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  carbo_sklearn_m_glm_pred[[i]] = df_pred_obs
  RMSE = RMSE(predictions,score_df_all_carbo_sklearn_m_test$response)
  carbo_sklearn_m_glm_RMSE[[i]] = RMSE
  print(RMSE)
  MAE = MAE(predictions,score_df_all_carbo_sklearn_m_test$response)
  carbo_sklearn_m_MAE[[i]] = MAE
  df_pred_obs = cbind(predictions,score_df_all_carbo_sklearn_m_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  colnames(df_pred_obs)[2] = "obs"
  carbo_sklearn_m_plot[[i]] = ggplot(df_pred_obs, aes(x = obs,y = predictions)) + geom_point() + geom_abline(intercept = 0, slope = 1)}
```


```{r  models doc sklearn_f}
xenome_human_mouse_RSEM_sklearn_f_doc$xeno = rownames(xenome_human_mouse_RSEM_sklearn_f_doc)
score_df_all_doc_sklearn_f = merge(xenome_human_mouse_RSEM_sklearn_f_doc,doc20_quant_score, by = "xeno")
rownames(score_df_all_doc_sklearn_f) = score_df_all_doc_sklearn_f$xeno
score_df_all_doc_sklearn_f$xeno = NULL
colnames(score_df_all_doc_sklearn_f)[16] = "response"

split_10_doc_sklearn_f = read.table("input/split_10_doc20_quant.txt")
doc_sklearn_f_glm = list()
doc_sklearn_f_glm_pred = list()
doc_sklearn_f_glm_RMSE = list()
doc_sklearn_f_MAE = list()
doc_sklearn_f_plot = list()

for (i in 1:10){
  split = split_10_doc_sklearn_f[,i]
  score_df_all_doc_sklearn_f_train = score_df_all_doc_sklearn_f[split,]
  score_df_all_doc_sklearn_f_test = score_df_all_doc_sklearn_f[-split,]
  glm_model = glm(response ~ ., data = score_df_all_doc_sklearn_f_train,maxit = 500)
  doc_sklearn_f_glm[[i]] = glm
  predictions <- glm_model %>% predict(score_df_all_doc_sklearn_f_test, type = "response")
  df_pred_obs = cbind(predictions,score_df_all_doc_sklearn_f_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  doc_sklearn_f_glm_pred[[i]] = df_pred_obs
  RMSE = RMSE(predictions,score_df_all_doc_sklearn_f_test$response)
  doc_sklearn_f_glm_RMSE[[i]] = RMSE
  print(RMSE)
  MAE = MAE(predictions,score_df_all_doc_sklearn_f_test$response)
  doc_sklearn_f_MAE[[i]] = MAE
  df_pred_obs = cbind(predictions,score_df_all_doc_sklearn_f_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  colnames(df_pred_obs)[2] = "obs"
  doc_sklearn_f_plot[[i]] = ggplot(df_pred_obs, aes(x = obs,y = predictions)) + geom_point() + geom_abline(intercept = 0, slope = 1)}
```
  
  
```{r  models doc sklearn_m}
#Can we predict these scores?
xenome_human_mouse_RSEM_sklearn_m_doc$xeno = rownames(xenome_human_mouse_RSEM_sklearn_m_doc)
score_df_all_doc_sklearn_m = merge(xenome_human_mouse_RSEM_sklearn_m_doc,doc20_quant_score, by = "xeno")
rownames(score_df_all_doc_sklearn_m) = score_df_all_doc_sklearn_m$xeno
score_df_all_doc_sklearn_m$xeno = NULL
colnames(score_df_all_doc_sklearn_m)[17] = "response"

split_10_doc_sklearn_m = read.table("input/split_10_doc20_quant.txt")

doc_sklearn_m_glm = list()
doc_sklearn_m_glm_pred = list()
doc_sklearn_m_glm_RMSE = list()
doc_sklearn_m_MAE = list()
doc_sklearn_m_plot = list()

for (i in 1:10){
  split = split_10_doc_sklearn_m[,i]
  score_df_all_doc_sklearn_m_train = score_df_all_doc_sklearn_m[split,]
  score_df_all_doc_sklearn_m_test = score_df_all_doc_sklearn_m[-split,]
  glm_model = glm(response ~ ., data = score_df_all_doc_sklearn_m_train,maxit = 500)
  doc_sklearn_m_glm[[i]] = glm
  predictions <- glm_model %>% predict(score_df_all_doc_sklearn_m_test, type = "response")
  df_pred_obs = cbind(predictions,score_df_all_doc_sklearn_m_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  doc_sklearn_m_glm_pred[[i]] = df_pred_obs
  RMSE = RMSE(predictions,score_df_all_doc_sklearn_m_test$response)
  doc_sklearn_m_glm_RMSE[[i]] = RMSE
  print(RMSE)
  MAE = MAE(predictions,score_df_all_doc_sklearn_m_test$response)
  doc_sklearn_m_MAE[[i]] = MAE
  df_pred_obs = cbind(predictions,score_df_all_doc_sklearn_m_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  colnames(df_pred_obs)[2] = "obs"
  doc_sklearn_m_plot[[i]] = ggplot(df_pred_obs, aes(x = obs,y = predictions)) + geom_point() + geom_abline(intercept = 0, slope = 1)}
```


#WGCNA hubs

```{r feature selection - WGCNA hubs }
hub_genes_carbo_ep_all_genes = read.table("input/hub_genes_carbo_ep_all_genes.txt")
hub_genes_carbo_ep_all_genes = as.character(hub_genes_carbo_ep_all_genes$x)
hubs_genes_human_carbo = hub_genes_carbo_ep_all_genes
hubs_genes_mouse_carbo_df = mouse_to_human_df[mouse_to_human_df$HGNC.symbol %in% hub_genes_carbo_ep_all_genes, ]
hubs_genes_mouse_carbo = as.character(hubs_genes_mouse_carbo_df$genes)

#No stromal epithelial hub genes

```

```{r feature selection - hubs doc }
hub_genes_doc20_ep_all_genes = read.table("input/hub_genes_doc20_ep_all_genes.txt")
hub_genes_doc20_ep_all_genes = as.character(hub_genes_doc20_ep_all_genes$x)
hub_genes_doc20_stromal_all_genes = read.table("input/hub_genes_doc20_stromal_all_genes.txt")
hub_genes_doc20_stromal_all_genes = as.character(hub_genes_doc20_stromal_all_genes$x)

hubs_doc_genes_stromal_human = mouse_to_human_df[mouse_to_human_df$genes %in% hub_genes_doc20_stromal_all_genes,]
hubs_doc_genes_stromal_human = as.character(hubs_doc_genes_stromal_human$HGNC.symbol)


hubs_genes_human_doc = c(hub_genes_doc20_ep_all_genes,hubs_doc_genes_stromal_human)
hubs_doc_genes_ep_mouse = mouse_to_human_df[mouse_to_human_df$HGNC.symbol %in%  hub_genes_doc20_ep_all_genes,]
hubs_genes_mouse_doc = c(as.character(hubs_doc_genes_ep_mouse$genes),hub_genes_doc20_stromal_all_genes)
```

```{r p-bulk hubs carbo}
xenome_human_RSEM_hubs = xenome_human_RSEM[,colnames(xenome_human_RSEM) %in% hubs_genes_human_carbo]
xenome_human_RSEM_hubs$pdx = rownames(xenome_human_RSEM_hubs)
xenome_human_RSEM_hubs_melt = melt(xenome_human_RSEM_hubs)

xenome_mouse_RSEM_hubs = xenome_mouse_RSEM[,colnames(xenome_mouse_RSEM) %in% hubs_genes_mouse_carbo]
#Change colnames to human homologs 
colnames_hubs_mouse = mouse_to_human_df[mouse_to_human_df$genes %in% hubs_genes_mouse_carbo,]
colnames_hubs_mouse_human = as.character(colnames_hubs_mouse$HGNC.symbol)
colnames_hubs_mouse$genes = as.character(colnames_hubs_mouse$genes)
hubs_both = intersect(colnames_hubs_mouse$genes,colnames(xenome_mouse_RSEM))
colnames_hubs_mouse = colnames_hubs_mouse[colnames_hubs_mouse$HGNC.symbol %in% hubs_genes_human_carbo,]
colnames_hubs_mouse = colnames_hubs_mouse[colnames_hubs_mouse$genes %in% hubs_both, ]
xenome_mouse_RSEM_hubs = xenome_mouse_RSEM[,colnames_hubs_mouse$genes]
colnames(xenome_mouse_RSEM_hubs) = colnames_hubs_mouse$HGNC.symbol


xenome_mouse_RSEM_hubs$pdx = rownames(xenome_mouse_RSEM_hubs)
xenome_mouse_RSEM_hubs_melt = melt(xenome_mouse_RSEM_hubs)
xenome_human_mouse_RSEM_hubs_melt = rbind(xenome_human_RSEM_hubs_melt,xenome_mouse_RSEM_hubs_melt)
xenome_human_mouse_RSEM_hubs_melt = aggregate(. ~pdx+variable, data=xenome_human_mouse_RSEM_hubs_melt, sum, na.rm=TRUE)
xenome_human_mouse_RSEM_hubs = recast(xenome_human_mouse_RSEM_hubs_melt, pdx ~ variable)
rownames(xenome_human_mouse_RSEM_hubs) = xenome_human_mouse_RSEM_hubs$pdx
xenome_human_mouse_RSEM_hubs$pdx = NULL 
```

```{r p-bulk hubs  doc}
xenome_human_RSEM_hubs_doc = xenome_human_RSEM[,colnames(xenome_human_RSEM) %in% hubs_genes_human_doc]
xenome_human_RSEM_hubs_doc$pdx = rownames(xenome_human_RSEM_hubs_doc)
xenome_human_RSEM_hubs_doc_melt = melt(xenome_human_RSEM_hubs_doc)

xenome_mouse_RSEM_hubs_doc = xenome_mouse_RSEM[,colnames(xenome_mouse_RSEM) %in% hubs_genes_mouse_doc]
#Change colnames to human homologs 
colnames_hubs_doc_mouse = mouse_to_human_df[mouse_to_human_df$genes %in% hubs_genes_mouse_doc,]
colnames_hubs_doc_mouse_human = as.character(colnames_hubs_doc_mouse$HGNC.symbol)
colnames_hubs_doc_mouse$genes = as.character(colnames_hubs_doc_mouse$genes)
xenome_mouse_RSEM_hubs_doc = xenome_mouse_RSEM_hubs_doc[,colnames_hubs_doc_mouse$genes]
colnames(xenome_mouse_RSEM_hubs_doc) = colnames_hubs_doc_mouse$HGNC.symbol


xenome_mouse_RSEM_hubs_doc$pdx = rownames(xenome_mouse_RSEM_hubs_doc)
xenome_mouse_RSEM_hubs_doc_melt = melt(xenome_mouse_RSEM_hubs_doc)
xenome_human_mouse_RSEM_hubs_doc_melt = rbind(xenome_human_RSEM_hubs_doc_melt,xenome_mouse_RSEM_hubs_doc_melt)
xenome_human_mouse_RSEM_hubs_doc_melt = aggregate(. ~pdx+variable, data=xenome_human_mouse_RSEM_hubs_doc_melt, sum, na.rm=TRUE)
xenome_human_mouse_RSEM_hubs_doc = recast(xenome_human_mouse_RSEM_hubs_doc_melt, pdx ~ variable)
rownames(xenome_human_mouse_RSEM_hubs_doc) = xenome_human_mouse_RSEM_hubs_doc$pdx
xenome_human_mouse_RSEM_hubs_doc$pdx = NULL 
#27
```


```{r  models carbo hubs}
#Can we predict these scores?
xenome_human_mouse_RSEM_hubs$xeno = rownames(xenome_human_mouse_RSEM_hubs)
score_df_all_carbo_hubs = merge(xenome_human_mouse_RSEM_hubs,carbo_quant_score, by = "xeno")
rownames(score_df_all_carbo_hubs) = score_df_all_carbo_hubs$xeno
score_df_all_carbo_hubs$xeno = NULL
colnames(score_df_all_carbo_hubs)[15] = "response"

split_10_carbo_hubs = read.table("input/split_10_carbo_quant.txt",sep = "\t")

carbo_hubs_glm = list()
carbo_hubs_glm_pred = list()
carbo_hubs_glm_RMSE = list()
carbo_hubs_MAE = list()
carbo_hubs_plot = list()

for (i in 1:10){
  split = split_10_carbo_hubs[,i]
  score_df_all_carbo_hubs_train = score_df_all_carbo_hubs[split,]
  score_df_all_carbo_hubs_test = score_df_all_carbo_hubs[-split,]
  glm_model = glm(response ~ ., data = score_df_all_carbo_hubs_train,maxit = 500)
  carbo_hubs_glm[[i]] = glm
  predictions <- glm_model %>% predict(score_df_all_carbo_hubs_test, type = "response")
  df_pred_obs = cbind(predictions,score_df_all_carbo_hubs_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  carbo_hubs_glm_pred[[i]] = df_pred_obs
  RMSE = RMSE(predictions,score_df_all_carbo_hubs_test$response)
  carbo_hubs_glm_RMSE[[i]] = RMSE
  print(RMSE)
  MAE = MAE(predictions,score_df_all_carbo_hubs_test$response)
  carbo_hubs_MAE[[i]] = MAE
  df_pred_obs = cbind(predictions,score_df_all_carbo_hubs_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  colnames(df_pred_obs)[2] = "obs"
  carbo_hubs_plot[[i]] = ggplot(df_pred_obs, aes(x = obs,y = predictions)) + geom_point() + geom_abline(intercept = 0, slope = 1)}
```
  
```{r models doc20 hubs }

xenome_human_mouse_RSEM_hubs_doc$xeno = rownames(xenome_human_mouse_RSEM_hubs_doc)
score_df_all_doc_hubs = merge(xenome_human_mouse_RSEM_hubs_doc,doc20_quant_score, by = "xeno")
rownames(score_df_all_doc_hubs) = score_df_all_doc_hubs$xeno
score_df_all_doc_hubs$xeno = NULL
colnames(score_df_all_doc_hubs)[28] = "response"

split_10_doc_hubs = read.table("input/split_10_doc20_quant.txt")

doc_hubs_glm = list()
doc_hubs_glm_pred = list()
doc_hubs_glm_RMSE = list()
doc_hubs_MAE = list()
doc_hubs_plot = list()

for (i in 1:10){
  split = split_10_doc_hubs[,i]
  score_df_all_doc_hubs_train = score_df_all_doc_hubs[split,]
  score_df_all_doc_hubs_test = score_df_all_doc_hubs[-split,]
  glm_model = glm(response ~ ., data = score_df_all_doc_hubs_train,maxit = 500)
  doc_hubs_glm[[i]] = glm
  predictions <- glm_model %>% predict(score_df_all_doc_hubs_test, type = "response")
  df_pred_obs = cbind(predictions,score_df_all_doc_hubs_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  doc_hubs_glm_pred[[i]] = df_pred_obs
  RMSE = RMSE(predictions,score_df_all_doc_hubs_test$response)
  doc_hubs_glm_RMSE[[i]] = RMSE
  print(RMSE)
  MAE = MAE(predictions,score_df_all_doc_hubs_test$response)
  doc_hubs_MAE[[i]] = MAE
  df_pred_obs = cbind(predictions,score_df_all_doc_hubs_test$response)
  df_pred_obs = data.frame(df_pred_obs)
  colnames(df_pred_obs)[2] = "obs"
   doc_hubs_plot[[i]] = ggplot(df_pred_obs, aes(x = obs,y = predictions)) + geom_point()+ geom_abline(intercept = 0, slope = 1)}
```


```{r plots}
png("RMSE_compare.png",1000,1000)
par(cex.axis=2)
par(cex.lab = 2)
par(cex.main = 2)
par(mar=c(12,5,2,2))

boxplot(unlist(ctd_carbo_RMSE),unlist(carbo_cor_glm_RMSE),unlist(carbo_rfe_glm_RMSE),unlist(carbo_hubs_glm_RMSE), unlist(ctd_doc_RMSE),unlist(doc_cor_glm_RMSE),unlist(doc_rfe_glm_RMSE),unlist(doc_hubs_glm_RMSE),names = c("ctd_carbo","cor_carbo","rfe_carbo","hubs_carbo","ctd_doc20","cor_doc20","rfe_doc20","hubs_doc"),ylab = "RMSE", main = "RMSE Comparison",las=2)  
dev.off()


RMSE_dataframe = data.frame(unlist(ctd_carbo_RMSE),unlist(carbo_cor_glm_RMSE),unlist(carbo_rfe_glm_RMSE),unlist(carbo_hubs_glm_RMSE),unlist(carbo_boruta_glm_RMSE), unlist(carbo_sklearn_f_glm_RMSE), unlist(carbo_sklearn_m_glm_RMSE), unlist(ctd_doc_RMSE),unlist(doc_cor_glm_RMSE),unlist(doc_rfe_glm_RMSE),unlist(doc_hubs_glm_RMSE),unlist(doc_boruta_glm_RMSE), unlist(doc_sklearn_f_glm_RMSE), unlist(doc_sklearn_m_glm_RMSE))


boxplot(unlist(ctd_carbo_RMSE),unlist(carbo_cor_glm_RMSE),unlist(carbo_rfe_glm_RMSE),unlist(carbo_hubs_glm_RMSE),unlist(carbo_boruta_glm_RMSE), unlist(carbo_sklearn_f_glm_RMSE), unlist(carbo_sklearn_m_glm_RMSE), unlist(ctd_doc_RMSE),unlist(doc_cor_glm_RMSE),unlist(doc_rfe_glm_RMSE),unlist(doc_hubs_glm_RMSE),unlist(doc_boruta_glm_RMSE), unlist(doc_sklearn_f_glm_RMSE), unlist(doc_sklearn_m_glm_RMSE),names = c("ctd_carbo","cor_carbo","rfe_carbo","hubs_carbo","boruta_carbo", "sklearn_carbo_f", "sklearn_carbo_m", "ctd_doc20","cor_doc20","rfe_doc20","hubs_doc", "boruta_doc", "sklearn_doc_f","sklearn_doc_m"),ylab = "RMSE", main = "RMSE Comparison",las=2)  


pdf("RMSE_compare.pdf")
par(cex.axis=1 )
par(cex.lab = 1)
par(cex.main = 1)
#par(mar=c(12,5,2,2))
boxplot(unlist(ctd_carbo_RMSE),unlist(carbo_cor_glm_RMSE),unlist(carbo_rfe_glm_RMSE),unlist(carbo_hubs_glm_RMSE),unlist(carbo_boruta_glm_RMSE), unlist(carbo_sklearn_f_glm_RMSE), unlist(carbo_sklearn_m_glm_RMSE), unlist(ctd_doc_RMSE),unlist(doc_cor_glm_RMSE),unlist(doc_rfe_glm_RMSE),unlist(doc_hubs_glm_RMSE),unlist(doc_boruta_glm_RMSE), unlist(doc_sklearn_f_glm_RMSE), unlist(doc_sklearn_m_glm_RMSE),names = c("ctd_carbo","cor_carbo","rfe_carbo","hubs_carbo","boruta_carbo", "sklearn_carbo_f", "sklearn_carbo_m", "ctd_doc20","cor_doc20","rfe_doc20","hubs_doc", "boruta_doc", "sklearn_doc_f","sklearn_doc_m"),ylab = "RMSE", main = "RMSE Comparison",las=2)  
dev.off()



pdf("MAE_compare.pdf")
par(cex.axis=2)
par(cex.lab = 2)
par(cex.main = 2)
par(mar=c(12,5,2,2))
boxplot(unlist(ctd_carbo_MAE),unlist(carbo_cor_MAE),unlist(carbo_rfe_MAE),unlist(carbo_hubs_MAE),unlist(carbo_boruta_MAE), unlist(carbo_sklearn_f_MAE), unlist(carbo_sklearn_m_MAE), unlist(ctd_doc_MAE),unlist(doc_cor_MAE),unlist(doc_rfe_MAE),unlist(doc_hubs_MAE),unlist(doc_boruta_MAE), unlist(doc_sklearn_f_MAE), unlist(doc_sklearn_m_MAE),names = c("ctd_carbo","cor_carbo","rfe_carbo","hubs_carbo","boruta_carbo", "sklearn_carbo_f", "sklearn_carbo_m", "ctd_doc20","cor_doc20","rfe_doc20","hubs_doc", "boruta_doc", "sklearn_doc_f","sklearn_doc_m"),ylab = "MAE", main = "MAE Comparison",las=2)  
dev.off()


pdf("MAE_compare.pdf")
par(cex.axis=1)
par(cex.lab = 1)
par(cex.main = 1)
boxplot(unlist(ctd_carbo_MAE),unlist(carbo_cor_MAE),unlist(carbo_rfe_MAE),unlist(carbo_hubs_MAE), unlist(ctd_doc_MAE),unlist(doc_cor_MAE),unlist(doc_rfe_MAE),unlist(doc_hubs_MAE),names = c("ctd_carbo","cor_carbo","rfe_carbo","hubs_carbo","ctd_doc20","cor_doc20","rfe_doc20","hubs_doc"),ylab = "MAE",main = "MAE Comparison",las=2)  
dev.off()


```

```{r ctd info genes only}
carbo_ep_info_genes = read.table("input/info_genes_carboep.txt")
carbo_ep_info_genes = as.character(carbo_ep_info_genes$x)
doc20_ep_info_genes = read.table("input/info_genes_doc20ep.txt")
doc20_ep_info_genes = as.character(doc20_ep_info_genes$x)

carbo_stromal_info_genes = read.table("input/info_genes_carbostromal.txt")
carbo_stromal_info_genes = as.character(carbo_stromal_info_genes$x)
#remove "E130307A14RIK" as it has no human homolog
carbo_stromal_info_genes = carbo_stromal_info_genes[!carbo_stromal_info_genes %in%  "E130307A14RIK"]

doc20_stromal_info_genes = read.table("input/info_genes_doc20stromal.txt")
doc20_stromal_info_genes = as.character(doc20_stromal_info_genes$x)
```

```{r which genes intersect}
ctd_genes_carbo = c(carbo_ep_info_genes,carbo_stromal_info_genes)
ctd_genes_doc = c(doc20_ep_info_genes,doc20_stromal_info_genes)

cor_genes_carbo = colnames(xenome_human_mouse_RSEM_cor)
cor_genes_carbo = cor_genes_carbo[! cor_genes_carbo %in% "xeno"]
cor_genes_doc = colnames(xenome_human_mouse_RSEM_cor_doc)
cor_genes_doc = cor_genes_doc[!cor_genes_doc %in% "xeno"]

rfe_genes_carbo = colnames(xenome_human_mouse_RSEM_rfe)
rfe_genes_carbo = rfe_genes_carbo[! rfe_genes_carbo %in% "xeno"]
rfe_genes_doc = colnames(xenome_human_mouse_RSEM_rfe_doc)
rfe_genes_doc = rfe_genes_doc[!rfe_genes_doc %in% "xeno"]

hubs_genes_carbo = colnames(xenome_human_mouse_RSEM_hubs)
hubs_genes_carbo = hubs_genes_carbo[! hubs_genes_carbo %in% "xeno"]
hubs_genes_doc = colnames(xenome_human_mouse_RSEM_hubs_doc)
hubs_genes_doc = hubs_genes_doc[!hubs_genes_doc %in% "xeno"]

boruta_genes_carbo = colnames(xenome_human_mouse_RSEM_bortua_carbo)
boruta_genes_carbo = boruta_genes_carbo[! boruta_genes_carbo %in% "xeno"]
boruta_genes_doc = colnames(xenome_human_mouse_RSEM_bortua_doc)
boruta_genes_doc = boruta_genes_doc[!boruta_genes_doc %in% "xeno"]


sklearn_f_genes_carbo = colnames(xenome_human_mouse_RSEM_sklearn_f_carbo)
sklearn_f_genes_carbo = sklearn_f_genes_carbo[! sklearn_f_genes_carbo %in% "xeno"]
sklearn_f_genes_doc = colnames(xenome_human_mouse_RSEM_sklearn_f_doc)
sklearn_f_genes_doc = sklearn_f_genes_doc[!sklearn_f_genes_doc %in% "xeno"]

sklearn_m_genes_carbo = colnames(xenome_human_mouse_RSEM_sklearn_m_carbo)
sklearn_m_genes_carbo = sklearn_m_genes_carbo[! sklearn_m_genes_carbo %in% "xeno"]
sklearn_m_genes_doc = colnames(xenome_human_mouse_RSEM_sklearn_m_doc)
sklearn_m_genes_doc = sklearn_m_genes_doc[!sklearn_m_genes_doc %in% "xeno"]



library(ComplexHeatmap)
#Too large for venn use upSetR
x_carbo = list(CTD = ctd_genes_carbo, Cor = cor_genes_carbo, RFE = rfe_genes_carbo,Hubs = hubs_genes_carbo,Boruta = boruta_genes_carbo,Select_KBest_f = sklearn_f_genes_carbo,Select_KBest_m =sklearn_m_genes_carbo)

x_carbo_matrix = list_to_matrix(x_carbo)

x_carbo_matrix_combo= make_comb_mat(x_carbo_matrix, mode = "intersect")

x_carbo_unlist = data.frame(table(unlist(x_carbo)))
x_carbo_unlist_2_more = x_carbo_unlist[x_carbo_unlist$Freq >= 2,]
write.table(x_carbo_unlist_2_more,"x_carbo_unlist_2_more.txt",sep = "\t")

pdf("upset_carbo.pdf")
UpSet(make_comb_mat(x_carbo_matrix, mode = "intersect"))
upset(x_carbo_matrix_combo, nset = 100,nintersects = NA, set_size.show = TRUE)
dev.off()

x_doc = list(CTD = ctd_genes_doc, Cor = cor_genes_doc, RFE = rfe_genes_doc,Hubs = hubs_genes_doc,Boruta = boruta_genes_doc,Select_KBest_f = sklearn_f_genes_doc,Select_KBest_m =sklearn_m_genes_doc)

x_doc_unlist = data.frame(table(unlist(x_doc)))
x_doc_unlist_2_more = x_doc_unlist[x_doc_unlist$Freq >= 2,]
write.table(x_doc_unlist_2_more,"x_doc_unlist_2_more.txt",sep = "\t")


pdf("upset_doc.pdf")
upset(fromList(x_doc), nset = 40)
dev.off()
```


